frontend/src/app/main/data/workspace/variants.cljs

```diff
@@ -0,0 +1,197 @@
+         (dwu/commit-undo-transaction undo-id))))))
+
+
+(defn remove-property
+  "Remove the variant property on the position pos
+   in all the components with this variant-id"
+  [variant-id pos]
+  (ptk/reify ::remove-property
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            data    (dsh/lookup-file-data state)
+            objects (-> (dsh/get-page data page-id)
+                        (get :objects))
+            related-components (find-related-components data objects variant-id)
+
+            changes (-> (pcb/empty-changes it page-id)
+                        (pcb/with-library-data data)
+                        (pcb/with-objects objects)
+                        (clv/generate-remove-property related-components pos))
+
+            undo-id (js/Symbol)]
+        (rx/of
+         (dwu/start-undo-transaction undo-id)
+         (dch/commit-changes changes)
+         (dwu/commit-undo-transaction undo-id))))))
+
+
+
+(defn add-new-property
+  "Add a new variant property to all the components with this variant-id"
+  [variant-id]
+  (ptk/reify ::add-new-property
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            data    (dsh/lookup-file-data state)
+            objects (-> (dsh/get-page data page-id)
+                        (get :objects))
+
+            related-components (find-related-components data objects variant-id)
+
+
+            property-name (str "Property" (-> related-components
+                                              first
+                                              :variant-properties
+                                              count
+                                              inc))
+
+            changes (-> (pcb/empty-changes it page-id)
+                        (pcb/with-library-data data)
+                        (pcb/with-objects objects)
+                        (clv/generate-add-new-property related-components property-name))
+
+
+            undo-id (js/Symbol)]
+        (rx/of
+         (dwu/start-undo-transaction undo-id)
+         (dch/commit-changes changes)
+         (dwu/commit-undo-transaction undo-id))))))
+
+(defn set-variant-id
+  "Sets the variant-id on a component"
+  [component-id variant-id]
+  (ptk/reify ::set-variant-id
+    ptk/WatchEvent
+    (watch [it state _]
+      (let [page-id (:current-page-id state)
+            data    (dsh/lookup-file-data state)
+            changes (-> (pcb/empty-changes it page-id)
+                        (pcb/with-library-data data)
+                        (pcb/update-component component-id #(assoc % :variant-id variant-id)))
+            undo-id  (js/Symbol)]
+        (rx/of
+         (dwu/start-undo-transaction undo-id)
+         (dch/commit-changes changes)
+         (dwu/commit-undo-transaction undo-id))))))
+
+(defn transform-in-variant
+  "Given the id of a main shape of a component, creates a variant structure for
+   that component"
+  [id]
```

Comment:
I'd call this directly `main-id` (or better `main-instance-id`). The operation `transform-in-variant` does not have one id that is obviously the chosen one. In this case you are giving the main instance id of the component, but for example you could have used the component id. So the name `id` in short is ambiguous, better to be more explicit.